File: easy/1122_relative_sort_array/solution.cpp
#include <map>
#include <vector>

class Solution {
public:
  std::vector<int> relativeSortArray(std::vector<int> &arr1,
                                     std::vector<int> &arr2) {
    std::map<int, int> arr1Counter;
    for (const int &num : arr1)
      ++arr1Counter[num];

    std::vector<int> sortedArr;

    for (const int &num : arr2) {
      for (int _ = 0; _ < arr1Counter[num]; ++_)
        sortedArr.push_back(num);
      arr1Counter[num] = 0;
    }

    for (const auto &[num, count] : arr1Counter) {
      if (count != 0) {
        for (int _ = 0; _ < count; ++_)
          sortedArr.push_back(num);
      }
    }

    return sortedArr;
  }
};

File: easy/1122_relative_sort_array/solution.py
from collections import Counter
from typing import List


class Solution:
    def relativeSortArray(self, arr1: List[int], arr2: List[int]) -> List[int]:
        arr1_counter = Counter(arr1)

        sorted_arr = []

        for num in arr2:
            sorted_arr.extend([num] * arr1_counter[num])
            arr1_counter.pop(num)

        for num, count in sorted(arr1_counter.items()):
            sorted_arr.extend([num] * count)

        return sorted_arr
